spot-client/src/common/backend/utils.js

```diff
@@ -0,0 +1,146 @@
+import { logger } from 'common/logger';
+import { getDeviceId } from 'common/utils/device-id';
+
+/**
+ * FIXME duplicated with post-to-endpoint and the jitter calculation in the loaders.
+ *
+ * Gets next timeout using the full jitter pattern.
+ *
+ * @param {number} retry - The retry number. It's 1 on the first retry.
+ * @returns {number} - The amount of waiting before trying another time given in milliseconds.
+ * @private
+ */
+function _getNextTimeout(retry) {
+    // 1st retry 0 - 2 seconds
+    // 2nd retry 0 - 4 seconds
+    // 3rd retry 0 - 16 seconds
+    return Math.floor(Math.random() * Math.pow(2, retry) * 1000);
+}
+
+/**
+ * Sends HTTP request using {@code fetch} with retries on network and server errors.
+ *
+ * @param {Object} fetchOptions - Options related to the fetch request.
+ * @param {string} fetchOptions.url - The URL to be passed to the fetch function.
+ * @param {Object} fetchOptions.requestOptions - The options to be passed to the fetch function.
+ * @param {string} fetchOptions.operationName - The name of the fetch operation that will appear in
+ * the retry related log entries.
+ * @param {number} retry - Keeps track of the retry attempt.
+ * @param {number} maxRetries - How many times will retry the request.
+ * @returns {Promise<Object>} - A promise resolved with the JSON parsed from the response.
+ */
+function fetchWithRetry(fetchOptions, retry = 0, maxRetries = 3) {
+    const { url, requestOptions, operationName } = fetchOptions;
+
+    return new Promise((resolve, reject) => {
+        fetch(url, requestOptions)
```

Comment:
I think fetch itself returns a promise so you dont have to wrap it in a promise.

Reply:
so then how to I `reject` or `resolve` it ?

Reply:
return Promise.reject() or throw an error. resolves should just happen if execution finishes without error.

Reply:
How would this part work ?
```
setTimeout(() => {
    fetchWithRetry(fetchOptions, _retry).then(resolve, reject);
}, timeout);
```

Reply:
Wrap only what needs to be wrapped in a promise instead of all the logic.
```
return new Promise((resolve, reject) => {
  setTimeout(() => {
      fetchWithRetry(fetchOptions, _retry).then(resolve, reject);
  }, timeout);
});

Reply:
I don't know... somehow for me one outer promise is easier to follow

Reply:
The rest of the blocks don't need to know about resolve being available in the scope; creating the promise around the timeout would limit the exposure of a variable to only where it's needed. You also don't have to be aware of the outer promise scope while reading through the chain and you won't have to keep context of where the resolve came from and if anything else uses it. Probably more to the point is it'd be easier to break out individual parts of the promise chain into separate replaceable, independent, testable pieces. This is why my preference would be to wrap only the setTimout in a promise. That said, I still approved the PR because I don't think it's important.